File: benchmarks-QuTiP/benchmarkutils.py
import timeit
import json
import qutip
import numpy as np

benchmark_directory = "benchmarks-QuTiP"
commitID = qutip.version.version
result_path = "../results/results-QuTiP-{}-{}.json"

def examplename(name):
    if name.endswith("]"):
        return name.rsplit("[", 1)[0]
    else:
        return name


def run_benchmark(f, *args, samples=5, evals=1):
    D = {"f": f, "args": args}
    t = timeit.repeat("f(*args)", globals=D, number=evals, repeat=samples)
    return min(t)/evals


def check(name, D, eps=1e-5):
    f = open("../checks/" + examplename(name) + ".json")
    data = json.load(f)
    for (N, result) in D.items():
        r = data[str(N)]
        if np.isnan(result) or abs(result-r)/abs(r) > eps:
            print("Warning: Result may be incorrect in", name, ": ", result, "<->", r)


def save(name, results):
    f = open(result_path.format(commitID, name), "w")
    json.dump(results, f)
    f.close()

File: benchmarks-QuantumOptics.jl/benchmarkutils.jl
module benchmarkutils

using QuantumOptics
using JSON

set_zero_subnormals(true)

# Detect the git commit hash
const rootpath = abspath(".")
const quantumoptics_directory = dirname(dirname(Base.functionloc(QuantumOptics.eval, Tuple{Void})[1]))
cd(quantumoptics_directory)
const commitID = readstring(`git rev-parse --verify HEAD`)[1:10]
cd(rootpath)
println("Detected QuantumOptics.jl version: ", commitID)

benchmark_directory = "benchmarks-QuantumOptics.jl"
checkvalues_directory = "checks"

function examplename(name)
    if endswith(name, "]")
        return rsplit(name, "[", limit=2)[1]
    end
    name
end

function check(name, D, eps=1e-5)
    check_path = "../checks/$(examplename(name)).json"
    if ispath(check_path)
        println("Checking against check file.")
        data = JSON.parsefile(check_path)
        for (N, result) in D
            r = data[string(N)]
            if isnan(result) || abs(result-r)/abs(r) > eps
                println("Warning: Result may be incorrect in", name, ": ", result, "<->", r)
            end
        end
    else
        println("No check file found - write results to check file.")
        f = open(check_path, "w")
        write(f, JSON.json(D))
        close(f)
    end
end

function save(name, results)
    result_path = "../results/results-QuantumOptics.jl-$commitID-$name.json"
    f = open(result_path, "w")
    write(f, JSON.json(results))
    close(f)
end

end # module